### PR TITLE
Update train_runtime reference number for clm gpt2_xl to fix pytest failure

### PR DESCRIPTION
**Pull request by hchauhan123**
With the new changes from [PR-1869](https://github.com/huggingface/optimum-habana/pull/1869) the pytest test_run_clm_gpt2-xl_deepspeed for clm gtp2_xl is failing due to train_runtime. 
`root:test_examples.py:785 train_runtime: le(236.54, 1.05 * 206.5775)`

Reference train_runtime = 216.90

The test should be updated with new reference train_runtime to avoid future failures. This is the new train_runtime results on G2 measured with PR-1869. Once PR-1869 is merged, this change should also be merged to avoid failures.

Command tested:
`RUN_SLOW=1 python -m pytest test_examples.py -v -s -k "test_run_clm_gpt2-xl_deepspeed"`

Results without PR-1869:
```
***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =      34.83
  memory_allocated (GB)       =      15.29
  total_flos                  = 41067876GF
  total_memory_available (GB) =      94.62
  train_loss                  =     2.7694
  train_runtime               = 0:03:23.79
  train_samples               =       2318
  train_samples_per_second    =     95.496
  train_steps_per_second      =      0.786
```
Note: The old train_runtime was 3:23 (~203 sec) which was within reference value.

Results with PR-1869:
```
***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =      37.55
  memory_allocated (GB)       =      15.29
  total_flos                  = 41067876GF
  total_memory_available (GB) =      94.62
  train_loss                  =     2.8371
  train_runtime               = 0:03:56.53
  train_samples               =       2318
  train_samples_per_second    =     95.361
  train_steps_per_second      =      0.785
```
Note: The new train_runtime is 03:56 (~236 sec) which should now be new reference value to fix the issue

Fixes # (issue)
Fixes the pytest test_run_clm_gpt2-xl_deepspeed

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by hchauhan123** [comment]
@yafshar  @mandy-li pls review.

**Comment by uartie** [comment]
LGTM

**Review by mandy-li — APPROVED** [review]
LGTM

**Review by yafshar — APPROVED** [review]
LGTM!

**Review by mounikamandava — APPROVED** [review]

**Comment by hchauhan123** [comment]
@regisss Can you please help to merge this?

**Comment by regisss** [comment]
@hchauhan123 The latest commits pushed to PR #1869 solve this issue and there is no runtime regression anymore. Can you try it and let me know if that works on your side please?

**Comment by hchauhan123** [comment]
@regisss I did a re-run with update PR-1869, and now I see the runtime is restored and there is no regression.
But I see you have updated the train runtime to even a better number of 191.6778 in the [commit](https://github.com/huggingface/optimum-habana/pull/1869/commits/4ce78490689efddd35149f9f0f80481b01cc2e8a#diff-fe92df0f65238cf7fc11c9e479336edaf6b1d2c656ea9bc20612d216d79588dbR105) . I don't see that better runtime at my end yet.

Below is my latest result on 1.20.0-543 with  transformers_future + PR-1869:
```
  ***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =      35.92
  memory_allocated (GB)       =      15.29
  total_flos                  = 41067876GF
  total_memory_available (GB) =      94.62
  train_loss                  =      2.844
  train_runtime               = 0:03:25.91
  train_samples               =       2318
  train_samples_per_second    =     95.636
  train_steps_per_second      =      0.787

```

And it fails with below error as the updated runtime in your commit is 191.6778
``` ERROR    root:test_examples.py:785 train_runtime: le(205.9134, 1.05 * 191.6778) ```

**Comment by regisss** [comment]
@hchauhan123 I just checked and I can reproduce this baseline. However, there are many other baselines that I can't reproduce on Gaudi3, probably due the setup of my CPU because compilation times seem to take longer.

**Comment by hchauhan123** [comment]
@regisss I did another run with a fresh setup of docker and now I am able to pass the test. My runtime of 198.07 was well within reference number and variable limit (191.6778 to 1.05*191.6778=201.2616). Did a couple of runs and it passed every time. 

Result:
```
***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =      34.83
  memory_allocated (GB)       =      15.29
  total_flos                  = 41067876GF
  total_memory_available (GB) =      94.62
  train_loss                  =      2.844
  train_runtime               = 0:03:18.07
  train_samples               =       2318
  train_samples_per_second    =     96.251
  train_steps_per_second      =      0.792
```

Since that the fix itself is in PR-1869, we don't need this PR anymore.

**Comment by hchauhan123** [comment]
@regisss 
Since the  [PR-1869](https://github.com/huggingface/optimum-habana/pull/1869) got merged and more commits were merged to transformers_future branch, I re-tested for this model and found train_runtime issue for Gaudi2 is still present as test is failing.

Current reference number for this model on Gaudi2:
train_runtime = 191.6778
upper limit for train_runtime to pass = 1.05 * 191.6778 = 201.261
Current status of transformers_future branch is at commit: 3e831a1

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/harshvar/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/harshvar/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">


OH branch | commit | corresponding PR   in OH | train_runtime   (sec) run1 | train_runtime   (sec) run2 | Date when test   was run
-- | -- | -- | -- | -- | --
transformers_future | 3e831a1 | 1698 | 242.16 (fail) | 237.11 (fail) | 4/4/2025
transformers_future | 4e12497 | 1869 | 200.55 (pass) | 206.52 (fail) | 4/4/2025
transformers_future | 4ce7849 | 1869 | 198.07 (pass) | 199.57 (pass) | 4/1/2025



</body>

</html>



My analysis:

(1) The first issue is with commit [3e831a1](https://github.com/huggingface/optimum-habana/commit/3e831a1bbda09a21bb9d708999788be2dc7c400e) This has increased the train_runtime and the regression is back due to this.
(2) When tested with commit 4e12497 which is [PR-1869](https://github.com/huggingface/optimum-habana/pull/1869) merged in transformers_future branch, the result is on borderline. Hence it passes sometimes and fails otherwise. This is because the reference value is too good 191.677 and the upper limit is up to 201.261 seconds.
(3) 3 days ago when PR-1869 was not merged and I had tested it with commit [4ce7849](https://github.com/huggingface/optimum-habana/pull/1869/commits/4ce78490689efddd35149f9f0f80481b01cc2e8a), I got train runtime around ~ 199 seconds, which I reported earlier. You mentioned you were always able to reproduce ~191.67 at that time. And I was still within the limit of 201.46 seconds and my test runs did not fail.

The original time for this model on Gaudi2 before [PR-1869](https://github.com/huggingface/optimum-habana/pull/1869) was 206.57. Post this PR, the train_runtime increased for multiple models and it was reflected in the reference values. Even the time on Gaudi3 for this model increased from 175 to 198. 

My understanding is that since this time variation is during warmup time, there is a fluctuation in train_runtime seen with [PR-1869](https://github.com/huggingface/optimum-habana/pull/1869). This can be reproduced once commit 3e831a1 is rolled back. With commit [3e831a1](https://github.com/huggingface/optimum-habana/commit/3e831a1bbda09a21bb9d708999788be2dc7c400e), the regression in train_runtime is a lot.

The reference train_runtime for this model on Gaudi2 needs an update. I don’t see how train_runtime becomes better from 206.57 before [PR-1869](https://github.com/huggingface/optimum-habana/pull/1869) to 191.677 after the PR while the train_runtime is increased on Gaudi3 from 175 to 198 second for this model.

**Comment by regisss** [comment]
@hchauhan123 https://github.com/huggingface/optimum-habana/commit/3e831a1bbda09a21bb9d708999788be2dc7c400e fixes an error that was introduced in those previous commits.

It seems that on Gaudi2 there is an extra compilation step during training that doesn't happen on Gaudi3. Which makes the total runtime larger, while there is no throughput regression (since the throughput calculation discards the first 3 training steps where compilations happen).

I'm not sure how much control we have on this, any idea? The change that triggers this behaviour is this line: https://github.com/huggingface/optimum-habana/blob/6aaef3a53f94a9c248033948604ab34d70fc0222/optimum/habana/transformers/trainer.py#L2614

**Comment by regisss** [comment]
@hchauhan123 I just pushed a new commit to update the baseline: https://github.com/huggingface/optimum-habana/pull/1698/commits/9025f8cc74bde36c2959da8b4bd93c40bc83af64

The thing is that an important fix to gradient accumulation was merged into Transformers, which is likely the cause for this extra compilation step (not sure why it doesn't happen on Gaudi3 though).

There is on regression on the throughput, which is more important IMO. Let's see if we understand better why this happens and then we can push a fix.

**Comment by hchauhan123** [comment]
@regisss Thanks for updating the train_runtime for the model on Gaudi2. But my pytest run still fails as the runtime you have updated is still too good. 
I get below numbers.
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/harshvar/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/harshvar/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


OH   branch | train_runtime   run1 | train_runtime   run2 | train_runtime   run3
-- | -- | -- | --
transformers_future | 237.317 | 241.093 | 238.51



</body>

</html>

Run-1:
```
***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =      34.83
  memory_allocated (GB)       =      15.29
  total_flos                  = 41067876GF
  total_memory_available (GB) =      94.62
  train_loss                  =     2.8371
  train_runtime               = 0:03:57.31
  train_samples               =       2318
  train_samples_per_second    =     95.574
  train_steps_per_second      =      0.787

ERROR    root:test_examples.py:785 train_runtime: le(237.3179, 1.05 * 224.3465)

```
Run-2
```
***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =      37.55
  memory_allocated (GB)       =      15.29
  total_flos                  = 41067876GF
  total_memory_available (GB) =      94.62
  train_loss                  =     2.8371
  train_runtime               = 0:04:01.09
  train_samples               =       2318
  train_samples_per_second    =     95.453
  train_steps_per_second      =      0.786


ERROR    root:test_examples.py:785 train_runtime: le(241.0939, 1.05 * 224.3465)
```

Run-3
```
***** train metrics *****
  epoch                       =        2.0
  max_memory_allocated (GB)   =      37.55
  memory_allocated (GB)       =      15.29
  total_flos                  = 41067876GF
  total_memory_available (GB) =      94.62
  train_loss                  =     2.8371
  train_runtime               = 0:03:58.51
  train_samples               =       2318
  train_samples_per_second    =     95.985
  train_steps_per_second      =       0.79

ERROR    root:test_examples.py:785 train_runtime: le(238.5103, 1.05 * 224.3465)

```

The tests will still fail in our QA with your runtime value 224.3465 (and upper limit is 1.05*224.34=235.56). Hence can you please merge this change with updated number. 
I agree there is no major performance drop. Regarding this extra compilation on G2, please feel free to discuss with Libin and maybe we can raise a jira.

**Review by regisss — APPROVED** [review]
Sounds good!
